### PR TITLE
drivers: meter: ade7816 : Add support for ADE7816

### PR DESCRIPTION
**Pull request by RaduSabau1**
## Pull Request Description

The ADE7816 is a highly accurate, multichannel metering
device that is capable of measuring one voltage channel and up
to six current channels. It measures line voltage and current and
calculates active and reactive energy, as well as instantaneous rms
voltage and current. The device incorporates seven sigma-delta
(Σ-Δ) ADCs with a high accuracy energy measurement core.
The six current input channels allow multiple loads to be measured
simultaneously. The voltage channel and the six current channels
each have a complete signal path allowing for a full range of
measurements.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Comment by RaduSabau1** [comment]
V3:
- Added driver documentation and API documentation.

**Comment by RaduSabau1** [comment]
V4:
- Fixed documentation failure

File: drivers/meter/ade7816/ade7816.h [hunk 1]
```diff
@@ -0,0 +1,498 @@
+/***************************************************************************//**
+ *   @file   ade7816.h
+ *   @brief  Header file of ADE7816 Driver.
+ *   @author Radu Sabau (radu.sabau@analog.com)
+********************************************************************************
+ * Copyright 2024(c) Analog Devices, Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (buha) [inline drivers/meter/ade7816/ade7816.h:10]: modify all licenses to the new 3-point BSD license you can find on `main`

File: drivers/meter/ade7816/ade7816.h [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+#include "no_os_util.h"
+#include "no_os_spi.h"
+#include "no_os_gpio.h"
+#include "no_os_irq.h"
+#include "no_os_error.h"
+#include "no_os_units.h"
```
Comment (buha) [inline drivers/meter/ade7816/ade7816.h:51]: this include is more fit in the .c file, it's not used in the .h

File: drivers/meter/ade7816/ade7816.c [hunk 3]
```diff
@@ -0,0 +1,1778 @@
...
+		if (ret)
+			goto remove_irq1_cb;
+
+		ret = no_os_irq_disable(init_param->irq_ctrl,
+					descriptor->gpio_irq1_desc->number);
+		if (ret)
+			goto remove_irq1_cb;
+
+		break;
+	case ADE7816_IRQ0_IRQ1:
```
Comment (buha) [inline drivers/meter/ade7816/ade7816.c:1507]: there is a lot of code repetition in these case switches, could you put the repetitive part after the switch and simplify a bit ?

File: drivers/meter/ade7816/README.rst [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+-----------------------
+
+Each interrupt available on the IRQ0 and IRQ1 pins depending on the STATUS0
+and STATUS1 registers can be configured using **ade7816_set_interrupt** API.
+
+Default interrupt handlers are provided within the driver in case user doesn't
+assign a callback to the initialization parameter referring to the specific IRQ
+pin.
+
+Priorities for the interrupt handlers can also be configured.
```
Comment (buha) [inline drivers/meter/ade7816/README.rst:77]: how ?

**Comment by RaduSabau1** [comment]
V5:
- Added new license header to driver and project files.
- Compacted the code in the initialization function.
- Addressed other comments.

**Review by buha — COMMENTED** [review]
i think the code in init can be compacted even more with some static helper functions, there is still some repetition remaining

**Comment by RaduSabau1** [comment]
V6:
- Added static helper function in order to compact ade7816_init code.

**Comment by RaduSabau1** [comment]
V8:
- Fix documentation CI failure.

**Review by buha — APPROVED** [review]

**Pull request merged**